File: backend/account-manager/src/main/java/com/jay/accountmanager/model/Bill.java
package com.jay.accountmanager.model;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.List;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.jay.accountmanager.dto.MonthAndDayDTO;
import com.jay.accountmanager.util.CustomDaysJsonConverter;
import com.jay.accountmanager.util.MonthAndDayJsonConverter;

import jakarta.persistence.Column;
import jakarta.persistence.Convert;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import lombok.Data;

@Data
@Entity
@Table(name = "bills")
public class Bill {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String name;

    @Column(nullable = false)
    private BigDecimal amount;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private BillType type;

    private String description;

    @Column
    private LocalDate dueDate;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private RecurrenceType recurrenceType;

    private Integer dayOfWeek;    // 1-7 for WEEKLY bills

    private Integer dayOfMonth;   // 1-31 for MONTHLY bills

    @Column(name = "month_and_day", columnDefinition = "TEXT")
    @Convert(converter = MonthAndDayJsonConverter.class)
    private MonthAndDayDTO monthAndDay;   // For YEARLY bills

    @Column(name = "custom_days", columnDefinition = "TEXT")
    @Convert(converter = CustomDaysJsonConverter.class)
    private List<Integer> customDays;    // Array of days for CUSTOM recurrence

    private LocalDate recurrenceEnd;

    @Column(nullable = false)
    private boolean isAutoPay;

    private String paymentMethod;

    private String category;

    @ManyToOne
    @JoinColumn(name = "person_id", nullable = false)
    @JsonBackReference
    private Person person;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "account_id")
    private Account account;
}

File: frontned/src/types/bill.ts
export enum BillType {
    RECURRING = 'RECURRING',
    HOME_LOAN = 'HOME_LOAN',
    EMI = 'EMI',
    UTILITIES = 'UTILITIES',
    GROCERIES = 'GROCERIES',
    ENTERTAINMENT = 'ENTERTAINMENT',
    HEALTHCARE = 'HEALTHCARE',
    EDUCATION = 'EDUCATION',
    TRANSPORTATION = 'TRANSPORTATION',
    OTHER = 'OTHER'
}

export enum RecurrenceType {
    DAILY = 'DAILY',
    WEEKLY = 'WEEKLY',
    MONTHLY = 'MONTHLY',
    YEARLY = 'YEARLY',
    ONE_TIME = 'ONE_TIME',
    CUSTOM = 'CUSTOM'
}

export interface Bill {
    id: number;
    name: string;
    amount: number;
    type: BillType;
    description?: string;
    recurrenceType: RecurrenceType;
    // Dynamic due date fields
    dueDate?: string; // For ONE_TIME bills
    dayOfWeek?: number; // 1-7 for WEEKLY bills
    dayOfMonth?: number; // 1-31 for MONTHLY bills
    monthAndDay?: { // For YEARLY bills
        month: number; // 1-12
        day: number; // 1-31
    };
    customDays?: number[]; // Array of days for CUSTOM recurrence
    recurrenceEnd?: string;
    autoPay: boolean;
    accountId?: number;
    paymentMethod?: string;
    category?: string;
}

export interface BillRequest {
    name: string;
    amount: number;
    type: BillType;
    description?: string;
    recurrenceType: RecurrenceType;
    // Dynamic due date fields
    dueDate?: string;
    dayOfWeek?: number;
    dayOfMonth?: number;
    monthAndDay?: {
        month: number;
        day: number;
    };
    customDays?: number[];
    recurrenceEnd?: string;
    autoPay: boolean;
    accountId?: number;
    paymentMethod?: string;
    category?: string;
}
